core/command_line_parser.h

```diff
@@ -0,0 +1,246 @@
+
+	// Helpers for the function above that parse a specific case
+	int _validate_positional(const String &p_arg, int p_current_idx);
+	int _validate_adjacent(const String &p_arg, const String &p_prefix, int p_separator);
+	int _validate_short(const String &p_arg, const String &p_prefix, int p_current_idx);
+	int _validate_long(const String &p_arg, const String &p_prefix, int p_current_idx);
+
+	// Validation helpers
+	const CommandLineOption *_validate_option(const String &p_name, const String &p_prefix);
+	int _validate_option_args(const CommandLineOption *p_option, const String &p_display_name, int p_current_idx, bool p_skip_first = false);
+	bool _validate_option_arg(const CommandLineOption *p_option, const String &p_display_name, const String &p_arg);
+
+	// Save information about parsed option
+	void _save_parsed_option(const CommandLineOption *p_option, const String &p_prefix, int p_idx, int p_args_count, const String &p_additional_value = String());
+	void _save_parsed_option(const CommandLineOption *p_option, const String &p_prefix, const String &p_value = String());
+	void _save_parsed_option(const CommandLineOption *p_option, int p_idx, int p_args_count);
+
+	// Help text printers
+	String _get_usage(const Vector<Pair<const CommandLineOption *, String>> &p_printable_options) const;
+	String _get_options_description(const OrderedHashMap<String, PackedStringArray> &p_categories_data) const;
+
+	// Other utilies
+	String _to_string(const PackedStringArray &p_names) const; // Returns all option names separated by commas with all prefix variants.
+	String _get_prefixed_longest_name(const PackedStringArray &p_names) const; // Returns longest name with first available prefix (long or short).
+	ParsedPrefix _parse_prefix(const String &p_arg) const;
+	String _find_most_similar(const String &p_name) const;
+	static bool _contains_optional_options(const Vector<Pair<const CommandLineOption *, String>> &p_printable_options);
+
+protected:
+	static void _bind_methods();
+
+public:
+	Error parse_args(const PackedStringArray &p_args);
+	Error validate();
+
+	void add_option(const Ref<CommandLineOption> &p_option);
+	int get_options_count() const;
+	Ref<CommandLineOption> get_option(int p_idx) const;
+	void set_option(int p_idx, const Ref<CommandLineOption> &p_option);
+	void remove_option(int p_idx);
+	Ref<CommandLineOption> find_option(const String &p_name) const;
+
+	Ref<CommandLineOption> add_help_option();
+	Ref<CommandLineOption> add_version_option();
+
+	bool is_set(const Ref<CommandLineOption> &p_option) const;
+	String get_value(const Ref<CommandLineOption> &p_option) const;
+	PackedStringArray get_values(const Ref<CommandLineOption> &p_option) const;
+
+	String get_prefix(const Ref<CommandLineOption> &p_option) const;
+	PackedStringArray get_prefixes(const Ref<CommandLineOption> &p_option) const;
+
+	int get_count(const Ref<CommandLineOption> &p_option) const;
```

Comment:
I'm still not sure about this: https://github.com/godotengine/godot/pull/44594#discussion_r641928839

It's possible that this can be confused with `get_options_count()` while iterating options in the parser. I'd rename to:

```suggestion
	int get_occurence_count()(const Ref<CommandLineOption> &p_option) const;
```

Reply:
Okay, let's rename :)